src/QsCompiler/Core/ExpressionWalker.fs

```diff
@@ -0,0 +1,454 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
```

Comment:
I'm confused about the purpose of these walkers. Does the ExpressionTransformation.fs, StatementTransformation.fs, and TreeTransformation.fs do all of this already?

Reply:
The walker classes iterate through the syntax tree, without producing any new objects, while the transformation classes fold through the syntax tree, producing a new version of the tree (potentially sharing some structure with the source tree). For situations where you don’t need or want the new tree, the walker classes are slightly simpler and should run somewhat faster while producing less garbage in memory.

Reply:
Oh, yes I can see the utility in that. It is unfortunate that there must be this code duplication, but I don't see any reasonable way around that. At a future point (not needed for this PR) we may want to consider have a script automatically generate these walker classes from their associated transformation classes, but that would only really be useful if we expect them to change much.

Reply:
I spent some time trying to come up with a clever way to have one set of classes provide both functions, perhaps with some clever use of generics, but I wasn't clever enough to do so. :-(